akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,217 @@
+/**
+ *  Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.dispatch.affinity
+
+import java.util
+import java.util.concurrent._
+import java.util.concurrent.locks.{ Lock, ReentrantLock }
+import akka.dispatch.{ AbstractBoundedNodeQueue, DispatcherPrerequisites, ExecutorServiceConfigurator, ExecutorServiceFactory }
+import akka.dispatch.affinity.AffinityPool._
+import com.typesafe.config.Config
+import net.openhft.affinity.{ AffinityStrategies, AffinityThreadFactory }
+
+import scala.collection.mutable
+import scala.collection.JavaConversions._
+import scala.util.{ Failure, Try }
+
+class AffinityPool(numThreads: Int, affinityGroupSize: Int, tf: ThreadFactory) extends AbstractExecutorService {
+
+  // controls access to mutable such as list of workers, etc
+  private val mainLock = new ReentrantLock
+  @volatile private var poolState: PoolState = Running
+
+  // there is one queue for each thread
+  private val workQueues = Array.fill(numThreads)(new BoundedTaskQueue(affinityGroupSize))
+
+  //mutable set of all workers that are currently running
+  private val workers: mutable.Set[ThreadPoolWorker] = mutable.Set()
```

Comment:
Instead of the ReentrantLock + the mutable.Set, how about using https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentSkipListSet.html ?